### PR TITLE
[MemProf] Remove context id set from nodes and recompute on demand

### PR DESCRIPTION
**Pull request by teresajohnson**
The ContextIds set on the ContextNode struct is not technically needed
as we can compute it from either the callee or caller edge context ids.
Remove it and add a helper to recompute from the edges on demand. Also
add helpers to compute the node allocation type and whether the context
ids are empty from the edges without needing to first compute the node's
context id set, to minimize the runtime cost increase.

This yielded a 20% reduction in peak memory for a large thin link, for
about a 2% time increase (which is more than offset by some other recent
time efficiency improvements).

### CODE REVIEW

File: llvm/lib/Transforms/IPO/MemProfContextDisambiguation.cpp [hunk 1]
```diff
@@ -1103,6 +1154,57 @@ void CallsiteContextGraph<DerivedCCG, FuncTy, CallTy>::connectNewNode(
...
+#endif
+  // Node's context ids should be the union of both its callee and caller edge
+  // context ids.
+  if (Node->CallerEdges.size()) {
+    auto EI = Node->CallerEdges.begin();
+    auto &FirstEdge = *EI;
+    EI++;
+    DenseSet<uint32_t> CallerEdgeContextIds(FirstEdge->ContextIds);
+    for (; EI != Node->CallerEdges.end(); EI++) {
+      const auto &Edge = *EI;
```
Comment (kazutakahirata) [inline llvm/lib/Transforms/IPO/MemProfContextDisambiguation.cpp:1183]: May I suggest `drop_begin` like so?

```suggestion
    DenseSet<uint32_t> CallerEdgeContextIds(Node->CallerEdges.front()->ContextIds);
    for (const auto &Edge : llvm::drop_begin(Node->CallerEdges)) {
```

File: llvm/lib/Transforms/IPO/MemProfContextDisambiguation.cpp [hunk 2]
```diff
@@ -1103,6 +1154,57 @@ void CallsiteContextGraph<DerivedCCG, FuncTy, CallTy>::connectNewNode(
...
+    assert(NodeContextIds == CallerEdgeContextIds ||
+           set_is_subset(CallerEdgeContextIds, NodeContextIds));
+  }
+  if (Node->CalleeEdges.size()) {
+    auto EI = Node->CalleeEdges.begin();
+    auto &FirstEdge = *EI;
+    EI++;
+    DenseSet<uint32_t> CalleeEdgeContextIds(FirstEdge->ContextIds);
+    for (; EI != Node->CalleeEdges.end(); EI++) {
+      const auto &Edge = *EI;
```
Comment (kazutakahirata) [inline llvm/lib/Transforms/IPO/MemProfContextDisambiguation.cpp:1199]: Likewise, may I suggest `drop_begin` here?

```suggestion
    DenseSet<uint32_t> CalleeEdgeContextIds(Node->CalleeEdges.front()->ContextIds);
    for (const auto &Edge : llvm::drop_begin(Node->CalleeEdges)) {
```

**Review by kazutakahirata — APPROVED** [review]
LGTM modulo some comments.

File: llvm/lib/Transforms/IPO/MemProfContextDisambiguation.cpp [hunk 3]
```diff
@@ -262,8 +262,70 @@ class CallsiteContextGraph {
...
-    DenseSet<uint32_t> ContextIds;
+    // Get the list of edges from which we can compute allocation information
+    // such as the context ids and allocation type of this node.
+    const std::vector<std::shared_ptr<ContextEdge>> *
+    getEdgesWithAllocInfo() const {
+      // If node has any callees, compute from those, otherwise compute from
+      // callers (i.e. if this is the leaf allocation node).
+      if (!CalleeEdges.empty())
+        return &CalleeEdges;
+      else if (!CallerEdges.empty()) {
```
[the remark this answers is not in the document]
Comment (teresajohnson) [reply llvm/lib/Transforms/IPO/MemProfContextDisambiguation.cpp:273]: done

File: llvm/lib/Transforms/IPO/MemProfContextDisambiguation.cpp (hunk 1 above)
Comment (teresajohnson) [reply llvm/lib/Transforms/IPO/MemProfContextDisambiguation.cpp:1183]: done

File: llvm/lib/Transforms/IPO/MemProfContextDisambiguation.cpp (hunk 2 above)
Comment (teresajohnson) [reply llvm/lib/Transforms/IPO/MemProfContextDisambiguation.cpp:1199]: done

File: llvm/lib/Transforms/IPO/MemProfContextDisambiguation.cpp [hunk 4]
```diff
@@ -2446,13 +2515,6 @@ void CallsiteContextGraph<DerivedCCG, FuncTy, CallTy>::
   // Now perform some updates that are common to all cases: the NewCallee gets
   // the moved ids added, and we need to remove those ids from OldCallee and
   // update its alloc type (NewCallee alloc type updates handled above).
```
[the remark this answers is not in the document]
Comment (teresajohnson) [reply llvm/lib/Transforms/IPO/MemProfContextDisambiguation.cpp:2517]: Removed it completely. The parts talking about the node updates are no longer relevant, and I have a comment in the moved code about it needing to be after the edge updates.

**Pull request merged**